File: Mage/MageSplitViewController.m
//
//  MageSplitViewController.m
//  MAGE
//
//

#import "MageSplitViewController.h"
#import "UserUtility.h"
#import "MageSessionManager.h"
#import "MapViewController_iPad.h"
#import "FeedItemSelectionDelegate.h"
#import "Mage.h"
#import "MAGE-Swift.h"

@interface MageSplitViewController () <AttachmentSelectionDelegate, UserSelectionDelegate, ObservationSelectionDelegate, AttachmentViewDelegate, FeedItemSelectionDelegate, ObservationActionsDelegate>
@property(nonatomic, strong) UINavigationController *masterViewController;
@property(nonatomic, strong) UINavigationController *detailViewController;
@property(nonatomic, strong) MageSideBarController *sideBarController;
    @property(nonatomic, strong) MapViewController_iPad *mapViewController;
    @property(nonatomic, weak) UIBarButtonItem *masterViewButton;
    @property(nonatomic, strong) NSArray *mapCalloutDelegates;
@property (strong, nonatomic) NSMutableArray *childCoordinators;
@property (strong, nonatomic) AttachmentViewCoordinator *attachmentCoordinator;
@property (strong, nonatomic) id<MDCContainerScheming> scheme;

@end

@implementation MageSplitViewController

- (void) applyThemeWithContainerScheme:(id<MDCContainerScheming>)containerScheme {
    if (containerScheme != nil) {
        self.scheme = containerScheme;
    }

    self.masterViewController.navigationBar.barTintColor = containerScheme.colorScheme.primaryColorVariant;
    self.masterViewController.navigationBar.tintColor = containerScheme.colorScheme.onPrimaryColor;
    [self.masterViewController.navigationBar setTitleTextAttributes:
     @{NSForegroundColorAttributeName:containerScheme.colorScheme.onPrimaryColor}];

    self.masterViewController.navigationBar.prefersLargeTitles = NO;
    self.masterViewController.navigationItem.largeTitleDisplayMode = UINavigationItemLargeTitleDisplayModeNever;
    
    self.detailViewController.navigationBar.barTintColor = containerScheme.colorScheme.primaryColorVariant;
    self.detailViewController.navigationBar.tintColor = containerScheme.colorScheme.onPrimaryColor;
    [self.detailViewController.navigationBar setTitleTextAttributes:
     @{NSForegroundColorAttributeName:containerScheme.colorScheme.onPrimaryColor}];
    
    self.detailViewController.navigationBar.prefersLargeTitles = NO;
    self.detailViewController.navigationItem.largeTitleDisplayMode = UINavigationItemLargeTitleDisplayModeNever;
    
    self.view.backgroundColor = containerScheme.colorScheme.surfaceColor;
}

- (instancetype) initWithScheme: (id<MDCContainerScheming>) containerScheme {
    self = [super init];
    self.scheme = containerScheme;
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setMaximumPrimaryColumnWidth:426];
    [self setPreferredPrimaryColumnWidthFraction:1.0];
    self.preferredDisplayMode = UISplitViewControllerDisplayModeAllVisible;
    self.childCoordinators = [[NSMutableArray alloc] init];
    
    [[Mage singleton] startServicesAsInitial:YES];
    
    self.delegate = self;
        
    self.sideBarController = [[MageSideBarController alloc] initWithContainerScheme:self.scheme];
    self.sideBarController.delegate = self;
    self.masterViewController = [[UINavigationController alloc] initWithRootViewController:self.sideBarController];
    
    self.mapViewController = [[MapViewController_iPad alloc] initWithScheme:self.scheme];
    self.detailViewController = [[UINavigationController alloc] initWithRootViewController:self.mapViewController];
    
    self.viewControllers = [NSArray arrayWithObjects:self.masterViewController, self.detailViewController, nil];

    self.mapViewController.mapDelegate.mapCalloutDelegate = self.mapViewController;
    
    [self applyThemeWithContainerScheme:self.scheme];
}

- (void) viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];

    self.masterViewButton = self.displayModeButtonItem;
    
    if(!UIWindow.isLandscape) {
        [self ensureButtonVisible];
    }
}

- (void)userDetailSelected:(User *) user {
    [[UIApplication sharedApplication] sendAction:self.masterViewButton.action to:self.masterViewButton.target from:nil forEvent:nil];
    [self.mapViewController userDetailSelected:user];
}

- (void)selectedUser:(User *) user {
    [self.mapViewController selectedUser:user];
}

- (void)selectedUser:(User *) user region:(MKCoordinateRegion) region {
    [self.mapViewController selectedUser:user region:region];
}

- (void)selectedObservation:(Observation *) observation {
    [self.mapViewController selectedObservation:observation];
}

- (void)selectedObservation:(Observation *) observation region:(MKCoordinateRegion) region {
    [self.mapViewController selectedObservation:observation region:region];
}

- (void)observationDetailSelected:(Observation *)observation {
    [[UIApplication sharedApplication] sendAction:self.masterViewButton.action to:self.masterViewButton.target from:nil forEvent:nil];
    [self.mapViewController observationDetailSelected:observation];
}

- (void) feedItemSelected:(FeedItem *)feedItem {
    [[UIApplication sharedApplication] sendAction:self.masterViewButton.action to:self.masterViewButton.target from:nil forEvent:nil];
    [self.mapViewController feedItemSelected:feedItem];
}

- (void) ensureButtonVisible {
    self.masterViewButton.title = self.sideBarController.title;
    self.masterViewButton.style = UIBarButtonItemStylePlain;
    
    self.mapViewController.navigationItem.leftBarButtonItem = self.masterViewButton;
}

- (void)splitViewController:(UISplitViewController *)svc willChangeToDisplayMode:(UISplitViewControllerDisplayMode)displayMode {
    self.masterViewButton = svc.displayModeButtonItem;
    if (displayMode == UISplitViewControllerDisplayModePrimaryOverlay) {
        [self ensureButtonVisible];
    } else if (displayMode == UISplitViewControllerDisplayModePrimaryHidden) {
        [self ensureButtonVisible];
    } else if (displayMode == UISplitViewControllerDisplayModeAllVisible) {
        self.mapViewController.navigationItem.leftBarButtonItem = nil;
    }
}

- (void) doneViewingWithCoordinator:(NSObject *)coordinator {
    [self.childCoordinators removeObject:coordinator];
    self.attachmentCoordinator = nil;
}

- (void) selectedAttachment:(Attachment *)attachment {
    if (self.attachmentCoordinator) {
        [self.attachmentCoordinator setAttachmentWithAttachment:attachment];
    } else {
        self.attachmentCoordinator = [[AttachmentViewCoordinator alloc] initWithRootViewController:self.mapViewController.navigationController attachment:attachment delegate:self scheme:_scheme];
        [self.childCoordinators addObject:self.attachmentCoordinator];
        [self.attachmentCoordinator start];
    }
}

- (void)selectedUnsentAttachment:(NSDictionary *)unsentAttachment {
    self.attachmentCoordinator = [[AttachmentViewCoordinator alloc] initWithRootViewController:self.mapViewController.navigationController url:unsentAttachment[@"localPath"] contentType: unsentAttachment[@"contentType"] delegate:self scheme:_scheme];
    [self.attachmentCoordinator start];
}


@end

File: Mage/MageSplitViewController.swift
//
//  MageSplitViewController.swift
//  MAGE
//
//  Created by Daniel Barela on 9/7/21.
//  Copyright © 2021 National Geospatial Intelligence Agency. All rights reserved.
//

import Foundation

class MageSplitViewController : UISplitViewController {
    @Injected(\.attachmentRepository)
    var attachmentRepository: AttachmentRepository
    
    @Injected(\.observationRepository)
    var observationRepository: ObservationRepository
    
    var bottomSheet: MDCBottomSheetController?
    
    var startStraightLineNavigationObserver: AnyObject?

    var scheme: MDCContainerScheming?;
    var masterViewController: UINavigationController?;
    var detailViewController: UINavigationController?;
    var sideBarController: MageSideBarController?;
    var mapViewController: MapViewController_iPad?;
    var masterViewButton: UIBarButtonItem?;
    var mapCalloutDelegates: [Any] = [];
    var childCoordinators: [NSObject] = [];
    var attachmentCoordinator: AttachmentViewCoordinator?;
    
    var router = MageRouter()
    
    init(frame: CGRect) {
        super.init(nibName: nil, bundle: nil);
    }
    
    required init(coder aDecoder: NSCoder) {
        fatalError("This class does not support NSCoding")
    }
    
    @objc convenience public init(containerScheme: MDCContainerScheming) {
        self.init(frame: CGRect.zero);
        self.scheme = containerScheme;
    }
    
    func applyTheme(withContainerScheme containerScheme: MDCContainerScheming?) {
        guard let containerScheme = containerScheme else {
            return
        }

        self.scheme = containerScheme;
        
        self.view.backgroundColor = containerScheme.colorScheme.surfaceColor;
    }
    
    override func viewDidLoad() {
        super.viewDidLoad();
        self.maximumPrimaryColumnWidth = 426;
        self.preferredPrimaryColumnWidthFraction = 1.0;
        self.preferredDisplayMode = .oneBesideSecondary;
        
        Mage.singleton.startServices(initial: true);
        
        self.delegate = self;
        
        self.sideBarController = MageSideBarController(scheme: self.scheme);
        self.masterViewController = UINavigationController(rootViewController: self.sideBarController!);

        self.mapViewController = MapViewController_iPad(delegate: nil, scheme: self.scheme!);
        self.detailViewController = UINavigationController(rootViewController: self.mapViewController!)

        self.viewControllers = [self.masterViewController!, self.detailViewController!]
        
        self.applyTheme(withContainerScheme: self.scheme);
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated);
        
        self.masterViewButton = self.displayModeButtonItem;
        
        if (!UIWindow.isLandscape) {
            ensureButtonVisible();
        }
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated);
        
        startStraightLineNavigationObserver = NotificationCenter.default.addObserver(forName: .StartStraightLineNavigation, object: nil, queue: .main) { notification in
            guard let _: StraightLineNavigationNotification = notification.object as? StraightLineNavigationNotification else {
                return;
            }
            self.detailViewController?.popToRootViewController(animated: false);
        }
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated);
        
        if let startStraightLineNavigationObserver = startStraightLineNavigationObserver {
            NotificationCenter.default.removeObserver(startStraightLineNavigationObserver)
        }
        startStraightLineNavigationObserver = nil
    }
    
    func ensureButtonVisible() {
        self.masterViewButton?.title = self.sideBarController?.title;
        self.masterViewButton?.style = .plain;
        self.mapViewController?.navigationItem.leftBarButtonItem = self.masterViewButton;
    }
    
    
    
}

extension MageSplitViewController: UISplitViewControllerDelegate {
    func splitViewController(_ svc: UISplitViewController, willChangeTo displayMode: UISplitViewController.DisplayMode) {
        self.masterViewButton = svc.displayModeButtonItem;
        if (displayMode == .oneOverSecondary) {
            ensureButtonVisible();
        } else if (displayMode == .secondaryOnly) {
            ensureButtonVisible();
        } else if (displayMode == .oneBesideSecondary) {
            ensureButtonVisible();
        }
    }
}
